framework/src/vectorpostprocessors/MaterialVectorPostprocessor.C

```diff
@@ -0,0 +1,109 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "MaterialVectorPostprocessor.h"
+#include "Material.h"
+#include "IndirectSort.h"
+
+#include "libmesh/quadrature.h"
+
+template<>
+InputParameters validParams<MaterialVectorPostprocessor>()
+{
+  InputParameters params = validParams<ElementVectorPostprocessor>();
+  params.addRequiredParam<std::string>("material", "Material for which all properties will be recorded.");
+  params.addParam<std::vector<unsigned int>>("elem_ids", std::vector<unsigned int>(), "Element IDs to print data for (others are ignored). Default is all elements.");
+  return params;
+}
+
+MaterialVectorPostprocessor::MaterialVectorPostprocessor(const InputParameters & parameters) :
+    ElementVectorPostprocessor(parameters),
+    _elem_filt_ids(getParam<std::vector<unsigned int>>("elem_ids")),
+    _elem_filter(_elem_filt_ids.begin(), _elem_filt_ids.end()),
+    _elem_ids(declareVector("elem_id")),
+    _qp_ids(declareVector("qp_id"))
+{
+  auto & mat = getMaterialByName(getParam<std::string>("material"), true);
+  auto & prop_names = mat.getSuppliedItems();
+  for (auto & prop : prop_names)
+  {
+    _prop_vecs.push_back(&declareVector(prop));
+    _prop_refs.push_back(&getMaterialProperty<Real>(prop));
```

Comment:
These vectors accumulate entries only n_quadrature_points at a time - which will generally be small compared to the total size of the vectors (after looping over multiple elements).  Since the vector auto-resizing algorithm is exponential - a resize buys us negligible performance improvement (actually might slow things down with extra function calls).